File: clients/ios/RemoteEverything/App/AppModel.swift
import Combine
import Foundation
import SwiftUI
import UIKit

/// The app's single coordinator: coordinates state across Controllers and exposes
/// unified state for SwiftUI views.
@MainActor
final class AppModel: ObservableObject {

    // MARK: - Published UI state

    @Published private(set) var settings: ClientSettings
    @Published var notice: Notice?
    @Published var path: [AppRoute] = []
    @Published var selectedNodeID: String?
    /// What the wide (two-column) layout's right pane shows instead of a node:
    /// settings or pairing, embedded beside the list the way Android embeds
    /// them (084aa92). On a phone these are a sheet and a pushed page instead.
    @Published var pane: AppPane?
    @Published private(set) var updateState: UpdateState = .idle

    /// The two panes a wide screen may hold beside the node list.
    enum AppPane: Equatable {
        case settings
        case addNode
    }

    // S2 View Bindings (delegated to pairingSession)
    var addNodeText: String {
        get { pairingSession.addNodeText }
        set { pairingSession.addNodeText = newValue }
    }
    var deviceNameDraft: String {
        get { pairingSession.deviceNameDraft }
        set { pairingSession.deviceNameDraft = newValue }
    }

    // MARK: - Controllers

    let nodesController: NodesController
    let catalogController: CatalogController
    let pairingSession: PairingSession

    // MARK: - Collaborators

    private let store: ClientStore
    private let vault: IdentityVault
    private let pairingCoordinator: PairingCoordinator
    private let clients: GatewayClientPool
    private let pathSelector = PathSelector.Cache()
    private let network = NetworkMonitor()
    private let updateChecker: UpdateChecker

    private var refreshLoop: Task<Void, Never>?
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Forwarded Properties for Views

    var identities: [Identity] { nodesController.identities }
    var nodes: [Node] { nodesController.nodes }
    var stagedSetups: [StagedSetup] { nodesController.stagedSetups }
    var identityConditions: [String: IdentityCondition] { nodesController.identityConditions }
    var isRefreshing: Bool { nodesController.isRefreshing }
    var catalogs: [String: CatalogState] { catalogController.catalogs }
    var pairingPhase: PairingPhase { pairingSession.pairingPhase }
    var invitation: SetupURI.Invitation? { pairingSession.invitation }
    var invitationError: SetupURI.Rejected? { pairingSession.invitationError }

    init(
        store: ClientStore = ClientStore(),
        vault: IdentityVault = IdentityVault(),
        updateChecker: UpdateChecker = UpdateChecker(),
        clients: GatewayClientPool = GatewayClientPool()
    ) {
        self.store = store
        self.vault = vault
        self.updateChecker = updateChecker
        self.clients = clients
        self.settings = store.loadSettings()

        let stages = StagedSetupStore(directory: store.pairingDirectory)
        let pairingCoord = PairingCoordinator(vault: vault, stages: stages) { origin, credential, pin in
            if credential == nil {
                return clients.pairingClient(origin: origin, pin: pin)
            }
            return GatewayClient(origin: origin, credential: credential, pin: pin)
        }
        self.pairingCoordinator = pairingCoord

        let nodesCtrl = NodesController(
            store: store,
            vault: vault,
            clients: clients,
            pathSelector: pathSelector,
            network: network
        )
        self.nodesController = nodesCtrl

        let catalogCtrl = CatalogController(
            nodesController: nodesCtrl,
            clients: clients,
            vault: vault,
            pathSelector: pathSelector
        )
        self.catalogController = catalogCtrl

        let pairingSess = PairingSession(
            pairing: pairingCoord,
            nodesController: nodesCtrl,
            clients: clients
        )
        self.pairingSession = pairingSess

        // Wire event handlers between controllers
        catalogCtrl.onNotice = { [weak self] notice in
            self?.notice = notice
        }
        pairingSess.onNotice = { [weak self] notice in
            self?.notice = notice
        }
        pairingSess.onCatalogReady = { [weak self] nodeID, catalog in
            self?.catalogController.setCatalog(.ready(catalog.applicationInfos), forNodeID: nodeID)
        }
        pairingSess.onClosePairPage = { [weak self] in
            self?.closePairPresentation()
        }

        // Propagate objectWillChange from child controllers to AppModel
        nodesCtrl.objectWillChange
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)

        catalogCtrl.objectWillChange
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)

        pairingSess.objectWillChange
            .sink { [weak self] _ in self?.objectWillChange.send() }
            .store(in: &cancellables)
    }

    // MARK: - Lifecycle

    func start() {
        Localization.apply(settings.language)
        // The window exists by now: apply the stored appearance at the level
        // that reaches every sheet and cover.
        applyAppearanceToWindows()
        network.onPathChange = { [weak self] in
            guard let self else { return }
            // The chosen path belongs to the network that is gone: it is retired
            // here so the next look re-decides between what is left, and whatever
            // node is on screen is asked again now rather than at the next tick
            // (Android networkChanged).
            self.pathSelector.reset()
            Task { await self.refreshNodes() }
            self.catalogController.recheckNow()
        }
        network.start()
        let (_, lost) = nodesController.reconcile(pairing: pairingCoordinator)
        if let first = lost.first {
            notice = .credentialLost(first)
        }
        Task { await refreshNodes() }
    }

    func scenePhaseChanged(_ phase: ScenePhase) {
        switch phase {
        case .active:
            startForegroundLoops()
            Task { await refreshNodes() }
        case .background:
            stopForegroundLoops()
        default:
            break
        }
    }

    private func startForegroundLoops() {
        startRefreshLoop()
        pairingSession.startPendingLoop()
    }

    private func stopForegroundLoops() {
        refreshLoop?.cancel()
        refreshLoop = nil
        pairingSession.stopLoops()
        catalogController.stopLoops()
    }

    private func startRefreshLoop() {
        refreshLoop?.cancel()
        refreshLoop = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                await self.refreshNodes()
                do {
                    try await Task.sleep(nanoseconds: Cadence.nodeRefreshNanos)
                } catch {
                    return
                }
            }
        }
    }

    // MARK: - Nodes

    func refreshNodes() async {
        // The minimum working-line time lives in the controller, next to
        // `isRefreshing`, so a fast refresh is still visible as the thing it is.
        await nodesController.refreshNodes()
    }

    func status(of node: Node) -> NodeStatus {
        nodesController.status(of: node)
    }

    func preferredPath(for node: Node) -> Path? { nodesController.preferredPath(for: node) }

    func node(withID id: String) -> Node? {
        nodesController.node(withID: id)
    }

    func identity(forPath path: Path) -> Identity? {
        nodesController.identity(forPath: path)
    }

    func nodeCount(forIdentity identity: Identity) -> Int {
        nodesController.nodeCount(forIdentity: identity)
    }

    // MARK: - Catalog & App Control

    func beginCatalogPolling(nodeID: String) {
        catalogController.beginCatalogPolling(nodeID: nodeID)
    }

    func endCatalogPolling() {
        catalogController.endCatalogPolling()
    }

    func loadCatalog(nodeID: String) async {
        await catalogController.loadCatalog(nodeID: nodeID)
    }

    func startApplication(nodeID: String, appID: String) {
        catalogController.startApplication(nodeID: nodeID, appID: appID)
    }

    func stopApplication(nodeID: String, appID: String) {
        catalogController.stopApplication(nodeID: nodeID, appID: appID)
    }

    func resolveWebTarget(nodeID: String, appID: String) async throws -> WebTarget {
        try await catalogController.resolveWebTarget(nodeID: nodeID, appID: appID)
    }

    // MARK: - S2 Pairing

    func beginAddNode() {
        pairingSession.beginAddNode()
        pushPairRoute()
    }

    /// Cancels the attempt in flight, keeping the invitation in the field.
    func cancelPairingAttempt() {
        pairingSession.cancelPairingAttempt()
    }

    func parseInvitation(_ text: String) {
        pairingSession.parseInvitation(text)
    }

    func handleIncomingURL(_ url: URL) {
        pairingSession.handleIncomingURL(url)
        addNodeText = url.absoluteString
        pushPairRoute()
    }

    func confirmPairing() async {
        await pairingSession.confirmPairing()
    }

    func resumeStagedSetups() async {
        await pairingSession.resumeStagedSetups()
    }

    // MARK: - The wide layout's panes

    /// Opens one of the wide screen's panes. Pairing starts a fresh flow; the
    /// detail column is cleared so the pane stands alone beside the list.
    func setPane(_ pane: AppPane?) {
        if pane == .addNode {
            pairingSession.beginAddNode()
            path.removeAll()
            selectedNodeID = nil
        }
        self.pane = pane
    }

    /// Opens settings: the pane on a wide screen, a pushed page on a phone — a
    /// sheet is not used, because a presented sheet cannot be relied on to
    /// follow an appearance change (the other clients' settings are a screen).
    func pushSettings() {
        guard !path.contains(where: { if case .settings = $0 { return true }; return false }) else { return }
        path.append(.settings)
    }

    /// Closes settings, whichever presentation it is: the pushed page on a
    /// phone, the pane on a wide screen.
    func closeSettings() {
        path.removeAll { if case .settings = $0 { return true }; return false }
        if pane == .settings { pane = nil }
    }

    /// Closes the pairing page, whichever presentation it is.
    func closePairPresentation() {
        popPairRoute()
        if pane == .addNode { pane = nil }
    }

    /// The pairing page is a pushed page on a phone and a pane on a wide
    /// screen; this is the pushed half, a sheet's worth of navigation.
    private func pushPairRoute() {
        guard !path.contains(where: { if case .pair = $0 { return true }; return false }) else { return }
        path.append(.pair)
    }

    private func popPairRoute() {
        path.removeAll { if case .pair = $0 { return true }; return false }
    }

    static func defaultDeviceName() -> String {
        PairingSession.defaultDeviceName()
    }

    // MARK: - S5 Settings

    var preferredColorScheme: ColorScheme? {
        switch settings.appearance {
        case .system: return nil
        case .light: return .light
        case .dark: return .dark
        }
    }

    func setLanguage(_ language: Language) {
        settings.language = language
        store.saveSettings(settings)
        // "Follow the system" is answered by what the system says *after* the
        // previous choice has stopped overriding it: persist first, resolve after.
        Localization.persistPreferredLanguage(language)
        Localization.apply(language)
    }

    func setAppearance(_ appearance: Appearance) {
        settings.appearance = appearance
        store.saveSettings(settings)
        applyAppearanceToWindows()
    }

    /// Appearance is applied at the window level, not through SwiftUI's
    /// preferredColorScheme: a window's overrideUserInterfaceStyle reaches
    /// every view the scene presents — the settings sheet included — through
    /// its trait collection, immediately and in both directions (including
    /// back to "follow the system"). preferredColorScheme alone left a
    /// presented sheet behind.
    func applyAppearanceToWindows() {
        let style: UIUserInterfaceStyle
        switch settings.appearance {
        case .light: style = .light
        case .dark: style = .dark
        case .system: style = .unspecified
        }
        for scene in UIApplication.shared.connectedScenes {
            guard let windowScene = scene as? UIWindowScene else { continue }
            for window in windowScene.windows {
                window.overrideUserInterfaceStyle = style
            }
        }
    }

    func forget(origin: String) async {
        guard identities.contains(where: { $0.origin == origin }) else { return }
        // A pairing staged against this gateway is part of what is being
        // forgotten: its row, its resume file and the credential it names go
        // with the identity (behavior README — forgetting does not leave a
        // half-pairing behind).
        pairingSession.discardStagedSetups(origin: origin)
        catalogController.clear()
        nodesController.forget(origin: origin)
        await refreshNodes()
    }

    func checkForUpdates() async {
        guard updateState != .checking else { return }
        updateState = .checking
        let result = await updateChecker.check(
            currentBuildNumber: AppVersion.buildNumber,
            currentProtocolVersion: AppVersion.protocolVersion
        )
        updateState = .result(result)
    }

    func dismissNotice(_ notice: Notice) {
        if self.notice == notice { self.notice = nil }
    }

    // --- connections & diagnostics -------------------------------------------

    func troubleFor(_ origin: String) -> LinkTrouble? {
        nodesController.troubleFor(origin)
    }

    /// The report a person copies out of a connection that will not answer.
    ///
    /// It is built here rather than on the screen because most of it is not on the
    /// screen: what the last few probes failed with, when this connection last
    /// answered, and which road each machine was reachable by.
    func troubleReport(identity: Identity) -> String {
        guard let trouble = nodesController.troubleFor(identity.origin) else { return "" }
        let say: (String, String?) -> String = { key, arg in
            if let arg {
                return l10n(key, arg)
            } else {
                return l10n(key)
            }
        }
        let machines = nodes
            .filter { node in node.paths.contains { $0.origin == identity.origin } }
            .map { node -> String in
                let roads = node.paths
                    .filter { $0.origin == identity.origin }
                    .map { path -> String in
                        let label = say(
                            path.link == .local ? MessageKeys.NODE_LAN : MessageKeys.NODE_TUNNEL,
                            nil
                        )
                        let mark = path.reachable == true ? "✓" : "✗"
                        let latency = (path.reachable == true && path.latencyMs != nil) ? " \(path.latencyMs!)ms" : ""
                        return "\(label) \(mark)\(latency)"
                    }
                    .joined(separator: " ")
                return "\(node.name) \(roads)"
            }

        let deviceModel = UIDevice.current.model
        let osVersion = UIDevice.current.systemVersion
        let system = "iOS \(osVersion) · \(deviceModel)"

        let formatter = DateFormatter()
        formatter.dateFormat = "MM-dd HH:mm:ss"

        return buildLinkReport(
            input: LinkReportInput(
                appName: say(MessageKeys.APP_NAME, nil),
                client: "\(AppVersion.versionName) (\(AppVersion.buildNumber))",
                protocolVersion: "\(AppVersion.protocolVersion)",
                system: system,
                origin: identity.origin,
                deviceName: identity.deviceName,
                network: network.describe(),
                machines: machines,
                trouble: trouble,
                attempts: nodesController.failedProbesFor(identity.origin),
                time: { formatter.string(from: $0) }
            ),
            say: say
        )
    }
}

File: clients/android/app/src/main/java/com/remoteeverything/app/AppModel.kt
package com.remoteeverything.app

import com.remoteeverything.app.i18n.LocaleHelper
import com.remoteeverything.app.i18n.l10n
import android.app.Application
import android.net.Network
import android.net.ConnectivityManager
import android.os.Build
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.remoteeverything.core.api.GatewayClientPool
import com.remoteeverything.core.diag.LinkReportInput
import com.remoteeverything.core.diag.LinkTrouble
import com.remoteeverything.core.diag.buildLinkReport
import com.remoteeverything.core.identity.AndroidKeyStoreIdentityVault
import com.remoteeverything.core.identity.IdentityVault
import com.remoteeverything.core.model.AppInfo
import com.remoteeverything.core.model.Cadence
import com.remoteeverything.core.model.ErrorCode
import com.remoteeverything.core.model.Identity
import com.remoteeverything.core.model.LinkKind
import com.remoteeverything.core.model.MessageKeys
import com.remoteeverything.core.model.Node
import com.remoteeverything.core.model.NodeStatus
import com.remoteeverything.core.model.Path
import com.remoteeverything.core.pairing.PairingService
import com.remoteeverything.core.pairing.SetupTransaction
import com.remoteeverything.core.pathselect.AndroidNetworkEnvironment
import com.remoteeverything.core.pathselect.NetworkEnvironment
import com.remoteeverything.core.setup.SetupUri
import com.remoteeverything.core.store.Appearance
import com.remoteeverything.core.store.Language
import com.remoteeverything.core.store.NodeCacheStore
import com.remoteeverything.core.store.Settings
import com.remoteeverything.core.store.SettingsStore
import com.remoteeverything.core.update.UpdateChecker
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.io.File

sealed interface PairingUiState {
    data object Idle : PairingUiState
    data class Working(val nodeName: String) : PairingUiState
    data class Pending(val nodeName: String) : PairingUiState
    data class Success(val nodeName: String) : PairingUiState
    data class Failed(val code: ErrorCode?, val key: String? = null) : PairingUiState
}

sealed interface CatalogUiState {
    data object Loading : CatalogUiState
    data class Ready(val apps: List<AppInfo>) : CatalogUiState
    data object Offline : CatalogUiState
    data object Unauthorized : CatalogUiState
    data class Failed(val code: ErrorCode?) : CatalogUiState
}

/** What a WebView needs to open one application: the origin, and what to pin it by. */
data class OpenTarget(val url: String, val serverPin: com.remoteeverything.core.model.ServerPin?)

data class AppUiState(
    val settings: Settings = Settings(),
    val identities: List<Identity> = emptyList(),
    val nodes: List<Node> = emptyList(),
    val refreshing: Boolean = false,
    val pairing: PairingUiState = PairingUiState.Idle,
    val catalog: CatalogUiState = CatalogUiState.Loading,
    val notice: Int? = null,
    /** Per connection origin, why it is not answering — empty for one that answers. */
    val trouble: Map<String, LinkTrouble> = emptyMap(),
)

sealed interface UpdateUiState {
    data object Idle : UpdateUiState
    data object Checking : UpdateUiState
    data object UpToDate : UpdateUiState
    data class Available(val versionName: String, val buildNumber: Int) : UpdateUiState
    data object ProtocolChanged : UpdateUiState
    data object Unreachable : UpdateUiState
}

/**
 * The app's single dispatcher and view model facade: coordinates the controllers,
 * holds global lifecycle, resolves web targets, and publishes consolidated UI state.
 */
class AppModel(application: Application) : AndroidViewModel(application) {

    private val repository = SettingsStore(application)
    private val vault: IdentityVault = AndroidKeyStoreIdentityVault()
    private val pool = GatewayClientPool()
    private val updateChecker = UpdateChecker()

    private val nodesController = NodesController(
        clientFactory = { identity -> pool.deviceClient(identity, vault) },
        cache = NodeCacheStore(File(application.filesDir, "nodes/cache.json")),
        // A phone with nothing to send on fails every connection the same way, and
        // saying so is more use than naming the first exception that came out.
        offline = { networkEnv.offline() },
    )

    private val pairingSession = PairingSession(
        scope = viewModelScope,
        pairingService = PairingService(
            vault = vault,
            transaction = SetupTransaction(File(application.filesDir, "pairing/staged.json")),
            repository = repository,
            clientFactory = { origin, pin, material ->
                if (material == null) pool.pairingClient(origin, pin)
                else com.remoteeverything.core.api.GatewayClients.device(origin, material, pin)
            },
        ),
        onNodesChanged = { refreshNodes() },
    )

    private val notice = MutableStateFlow<Int?>(null)

    private val catalogController = CatalogController(
        scope = viewModelScope,
        nodesController = nodesController,
        currentNetworkKey = ::currentNetworkKey,
        currentIdentities = { repository.currentIdentities() },
        // What one node looks like *now*: its paths carry the reachability of the
        // last probe, and a screen holding yesterday's answer would hold a dead
        // path (NodesController.refresh).
        currentNode = { id -> nodes.value.firstOrNull { node -> node.id == id } },
        onNotice = { notice.value = it },
    )

    private val nodes = MutableStateFlow<List<Node>>(emptyList())
    private val refreshing = MutableStateFlow(false)
    private var refreshJob: Job? = null
    private var refreshInFlight: Job? = null

    val state: StateFlow<AppUiState> = combine(
        repository.settings,
        repository.identities,
        nodes,
        refreshing,
        nodesController.trouble,
    ) { settings, identities, nodeList, isRefreshing, trouble ->
        AppUiState(
            settings = settings,
            identities = identities,
            nodes = nodeList,
            refreshing = isRefreshing,
            trouble = trouble,
        )
    }.combine(pairingSession.pairing) { current, pairingState -> current.copy(pairing = pairingState) }
        .combine(catalogController.catalog) { current, catalogState -> current.copy(catalog = catalogState) }
        .combine(notice) { current, message -> current.copy(notice = message) }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), AppUiState())

    val updateState = MutableStateFlow<UpdateUiState>(UpdateUiState.Idle)

    init {
        viewModelScope.launch { refreshNodes() }
        watchTheNetwork()
    }

    fun consumeNotice() {
        notice.value = null
    }

    /**
     * The phone changing networks is the phone asking a new question.
     *
     * A device that walks out of its Wi-Fi has not lost its machines: it has lost
     * one way to them, and the other way is a tunnel it is still holding. Waiting
     * for the next tick would leave the screen saying "offline" for a minute on a
     * phone that is on the internet the whole time, so a change of network is
     * answered at once: the list is read again and whatever node is on screen is
     * asked again, which is also how the choice of path gets recomputed.
     */
    private fun watchTheNetwork() {
        val manager = getApplication<Application>().getSystemService(ConnectivityManager::class.java)
            ?: return
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = networkChanged()
            override fun onLost(network: Network) = networkChanged()
            private fun networkChanged() = this@AppModel.networkChanged()
        }
        runCatching { manager.registerDefaultNetworkCallback(callback) }
            .onSuccess { networkCallback = callback }
    }

    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    fun networkChanged() {
        // The chosen path belongs to the network that is gone: it is retired here so
        // the next look re-decides between what is left.
        nodes.value.forEach { node -> nodesController.dropChoice(node.id) }
        refreshNodes(force = true)
        catalogController.recheckNow()
    }

    override fun onCleared() {
        networkCallback?.let { callback ->
            getApplication<Application>().getSystemService(ConnectivityManager::class.java)
                ?.unregisterNetworkCallback(callback)
        }
        super.onCleared()
    }

    fun refreshNodes(force: Boolean = false) {
        if (refreshing.value && !force) return
        // A forced refresh preempts the refresh in flight, not the foreground
        // loop: refreshJob is the periodic cadence, and canceling it would
        // leave the phone without its tick until it next comes to the front.
        if (force) refreshInFlight?.cancel()
        refreshing.value = true
        refreshInFlight = viewModelScope.launch {
            try {
                nodes.value = nodesController.refresh(repository.currentIdentities(), currentNetworkKey())
            } finally {
                refreshing.value = false
            }
        }
    }

    /** Started when the app comes to the front and stopped when it leaves. */
    fun startForegroundRefresh() {
        if (refreshJob?.isActive == true) return
        refreshJob = viewModelScope.launch {
            // Fresh on arrival, then on the cadence: the person who just came back
            // did not wait a minute to see what changed while they were away.
            refreshNodes()
            while (true) {
                delay(Cadence.nodeRefreshMs)
                refreshNodes()
            }
        }
    }

    fun stopForegroundRefresh() {
        refreshJob?.cancel()
        refreshJob = null
    }

    private val networkEnv = AndroidNetworkEnvironment(application)

    fun currentNetworkKey(): String = networkEnv.currentKey()

    /** The path a node is reached by right now, which is also what the row shows. */
    fun pathFor(node: Node): Path? = nodesController.pathFor(node, currentNetworkKey())

    /** Which gateway a node is reached through right now; null when none answers. */
    fun originFor(node: Node): String? = nodesController.originFor(node, currentNetworkKey())

    fun statusOf(node: Node): NodeStatus = nodesController.statusOf(node, currentNetworkKey())

    // --- pairing -------------------------------------------------------------

    /**
     * An invitation link waiting for the pair screen, as a one-shot event: the
     * deep link stages it, the navigation observes it, and the pair screen
     * consumes it. A plain field here was what re-opened the pair screen after
     * every recreation — language switches included — with the last link
     * filled in.
     */
    private val mutablePendingInvitation = MutableStateFlow<String?>(null)
    val pendingInvitation: StateFlow<String?> = mutablePendingInvitation

    fun stageInvitation(invitation: String) {
        mutablePendingInvitation.value = invitation
    }

    fun consumeInvitation() {
        mutablePendingInvitation.value = null
    }

    fun parseInvitation(text: String): SetupUri.Invitation? = pairingSession.parseInvitation(text)

    fun pair(invitation: SetupUri.Invitation, deviceName: String) = pairingSession.pair(invitation, deviceName)

    fun cancelPairing() = pairingSession.cancelPairing()

    fun resetPairing() = pairingSession.resetPairing()

    /** A waiting device, asked about now rather than at the next poll. */
    fun resumePendingPairing() = pairingSession.resumePendingPairing()

    // --- one node's applications --------------------------------------------

    fun watchNode(node: Node) = catalogController.watchNode(node)

    /**
     * A person asking again, from a screen that says something is unreachable: the
     * node list is read again — the paths it carries are what a catalog is fetched
     * through — and this node's catalog with it, now rather than at the next tick.
     */
    fun retryNode(node: Node) {
        refreshNodes()
        catalogController.watchNode(node, force = true)
    }

    fun stopWatching() = catalogController.stopWatching()

    fun control(node: Node, appId: String, start: Boolean) = catalogController.control(node, appId, start)

    /** Where the WebView is sent: the application's own origin, with the pin that origin must satisfy. */
    suspend fun open(node: Node, appId: String): OpenTarget? {
        val path = nodesController.choosePath(node, currentNetworkKey()) ?: return null
        val identity = repository.currentIdentities().firstOrNull { it.origin == path.origin } ?: return null
        val client = nodesController.clientFor(identity) ?: return null
        return try {
            val url = client.open(node.id, appId)
            OpenTarget(url = url, serverPin = identity.serverPin)
        } catch (e: Exception) {
            null
        }
    }

    // --- connections ---------------------------------------------------------

    /**
     * The report a person copies out of a connection that will not answer.
     *
     * It is built here rather than on the screen because most of it is not on the
     * screen: what the last few probes failed with, when this connection last
     * answered, and which road each machine was reachable by. The strings come from
     * the app's own locale (`LocaleHelper.wrap`), so the report reads the way the
     * rest of the app does.
     */
    fun troubleReport(identity: Identity): String {
        val trouble = nodesController.trouble.value[identity.origin] ?: return ""
        val context = LocaleHelper.wrap(getApplication())
        val say: (String, String?) -> String = { key, arg ->
            if (arg == null) l10n(context, key) else l10n(context, key, arg)
        }
        val machines = nodes.value
            .filter { node -> node.paths.any { it.origin == identity.origin } }
            .map { node ->
                val roads = node.paths
                    .filter { it.origin == identity.origin }
                    .joinToString(" ") { path ->
                        val label = say(
                            if (path.link == LinkKind.LOCAL) MessageKeys.NODE_LAN else MessageKeys.NODE_TUNNEL,
                            null,
                        )
                        val mark = if (path.reachable == true) "✓" else "✗"
                        val latency = path.latencyMs?.takeIf { path.reachable == true }?.let { " ${it}ms" }.orEmpty()
                        "$label $mark$latency"
                    }
                "${node.name} $roads"
            }
        return buildLinkReport(
            LinkReportInput(
                appName = say(MessageKeys.APP_NAME, null),
                client = "${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
                protocol = "${BuildConfig.PROTOCOL_VERSION}",
                system = "${Build.VERSION.RELEASE} (API ${Build.VERSION.SDK_INT}) · ${Build.MODEL}",
                origin = identity.origin,
                deviceName = identity.deviceName,
                network = networkEnv.describe(),
                machines = machines,
                trouble = trouble,
                attempts = nodesController.failedProbesFor(identity.origin),
                time = ::formatStamp,
            ),
            say,
        )
    }

    private fun formatStamp(at: Long): String =
        java.text.SimpleDateFormat("MM-dd HH:mm:ss", java.util.Locale.getDefault()).format(java.util.Date(at))

    fun forget(identity: Identity) {
        // The pairing still staged for this gateway is part of what is being
        // forgotten — its resume file and its approval polling go with the
        // identity (behavior README: forgetting leaves no half-pairing).
        pairingSession.forget(identity.origin)
        viewModelScope.launch {
            pool.drop(identity.origin)
            nodesController.drop(identity.origin)
            vault.delete(identity.origin)
            repository.saveIdentities(repository.currentIdentities().filterNot { it.origin == identity.origin })
            refreshNodes()
        }
    }

    // --- settings and updates ------------------------------------------------

    fun setLanguage(language: Language) {
        LocaleHelper.persistOverride(getApplication(), language)
        viewModelScope.launch { repository.setLanguage(language) }
    }

    fun setAppearance(appearance: Appearance) {
        viewModelScope.launch { repository.setAppearance(appearance) }
    }

    fun checkUpdates() {
        if (updateState.value == UpdateUiState.Checking) return
        updateState.value = UpdateUiState.Checking
        viewModelScope.launch {
            updateState.value = when (
                val result = updateChecker.check(BuildConfig.VERSION_CODE, BuildConfig.PROTOCOL_VERSION)
            ) {
                is UpdateChecker.Result.UpToDate -> UpdateUiState.UpToDate
                is UpdateChecker.Result.Available -> UpdateUiState.Available(result.versionName, result.buildNumber)
                is UpdateChecker.Result.ProtocolChanged -> UpdateUiState.ProtocolChanged
                is UpdateChecker.Result.Unreachable -> UpdateUiState.Unreachable
            }
        }
    }
}
